### PR TITLE
llama-cpp: fix blasSupport

### PR DESCRIPTION
**Pull request by ck3d**
## Description of changes

openblas was not provided as buildInput and added myself as maintainer

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/ll/llama-cpp/package.nix [hunk 1]
```diff
@@ -94,6 +96,7 @@ effectiveStdenv.mkDerivation (finalAttrs: {
     ++ optionals mpiSupport mpi
     ++ optionals openclSupport [ clblast ]
     ++ optionals rocmSupport rocmBuildInputs
+    ++ optional blasSupport openblas.dev
```
Comment (drupol) [inline pkgs/by-name/ll/llama-cpp/package.nix:99]: Let's use `optionals` here.

It's true, it's opinionated. But `optional` vs `optionals` is always an issue for newcomers.

`optional` implicitly wrap the value in a list, while `optionals` doesn't. `optionals` have a lower cognitive load than `optional`, and I think that we should use it here.

```suggestion
    ++ optionals blasSupport [ openblas.dev ]
```

Another thing: I also notice that the derivation `mpi` (line 96) should also be written `[ mpi ]`.

**Review by drupol — CHANGES_REQUESTED** [review]

File: pkgs/by-name/ll/llama-cpp/package.nix (hunk 1 above)
Comment (ck3d) [reply pkgs/by-name/ll/llama-cpp/package.nix]: thanks, I updated the PR

**Review by drupol — APPROVED** [review]

File: pkgs/by-name/ll/llama-cpp/package.nix [hunk 2]
```diff
@@ -91,9 +93,10 @@ effectiveStdenv.mkDerivation (finalAttrs: {
 
   buildInputs = optionals effectiveStdenv.isDarwin darwinBuildInputs
     ++ optionals cudaSupport cudaBuildInputs
-    ++ optionals mpiSupport mpi
+    ++ optionals mpiSupport [ mpi ]
     ++ optionals openclSupport [ clblast ]
     ++ optionals rocmSupport rocmBuildInputs
+    ++ optionals blasSupport [ openblas.dev ]
```
Comment (a-n-n-a-l-e-e) [inline pkgs/by-name/ll/llama-cpp/package.nix:99]:
```suggestion
    ++ optionals blasSupport [ openblas ]
```
why is this `dev` need to link to the libraries too.

Comment (ck3d) [reply pkgs/by-name/ll/llama-cpp/package.nix]: You are right, thanks. I updated this PR.

File: pkgs/by-name/ll/llama-cpp/package.nix [hunk 3]
```diff
@@ -150,7 +153,7 @@ effectiveStdenv.mkDerivation (finalAttrs: {
     homepage = "https://github.com/ggerganov/llama.cpp/";
     license = licenses.mit;
     mainProgram = "llama";
-    maintainers = with maintainers; [ dit7ya elohmeier philiptaron ];
+    maintainers = with maintainers; [ dit7ya elohmeier philiptaron ck3d ];
```
Comment (a-n-n-a-l-e-e) [inline pkgs/by-name/ll/llama-cpp/package.nix:156]: why do you want to be a maintainer? this is a pretty active packages but this looks like your first commit to the package.

Comment (ck3d) [reply pkgs/by-name/ll/llama-cpp/package.nix]: 1. I packaged local-ai, which depends on llama-cpp.
2. I saw breaking changes in the past, so I wanted to help.

Never mind, I removed myself from this list.

File: pkgs/by-name/ll/llama-cpp/package.nix [hunk 4]
```diff
@@ -16,6 +16,8 @@
 , clblast
 
 , blasSupport ? builtins.all (x: !x) [ cudaSupport metalSupport openclSupport rocmSupport vulkanSupport ]
+, openblas
```
Comment (NickCao) [inline pkgs/by-name/ll/llama-cpp/package.nix:19]: Or just `blas`?

Comment (ck3d) [reply pkgs/by-name/ll/llama-cpp/package.nix]: I added a second commit to switch from openblas to blas.

File: pkgs/by-name/ll/llama-cpp/package.nix (hunk 3 above)
Comment (drupol) [reply pkgs/by-name/ll/llama-cpp/package.nix]: IMHO, having new maintainers is fine.

Comment (philiptaron) [reply pkgs/by-name/ll/llama-cpp/package.nix]: > IMHO, having new maintainers is fine.

I agree.

File: pkgs/by-name/ll/llama-cpp/package.nix [hunk 5]
```diff
@@ -91,9 +93,10 @@ effectiveStdenv.mkDerivation (finalAttrs: {
 
   buildInputs = optionals effectiveStdenv.isDarwin darwinBuildInputs
     ++ optionals cudaSupport cudaBuildInputs
-    ++ optionals mpiSupport mpi
+    ++ optionals mpiSupport [ mpi ]
```
Comment (philiptaron) [inline pkgs/by-name/ll/llama-cpp/package.nix:96]: This change looks good (and supports @drupol's claim that `optional` vs. `optionals` distinction leads to errors.)

I only use `llama-cpp` with `cuda` and so don't have the understanding to experiment with the BLAS ecosystem.

**Review by philiptaron — APPROVED** [review]
I see that the CI on the corresponding backport passes.

**Comment by a-n-n-a-l-e-e** [comment]
strange that it was not failing in the compile without blas
https://github.com/ggerganov/llama.cpp/blob/b06c16ef9f81d84da520232c125d4d8a1d273736/CMakeLists.txt#L296
cause the cmake file looks for openblas64.pc and if that fails openblas.pc and should abort if not found. It did use to fail because i fixed it (added openblas64.pc) for ZHF 23.11.

anyway. the aarch64-darwin build fails.

**Comment by philiptaron** [comment]
> anyway. the aarch64-darwin build fails.

https://github.com/ggerganov/llama.cpp/pull/6118 I believe

**Pull request merged**